File: Tree/secondmin.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int findSecondMinHelper(struct TreeNode* root) {
    int smin;
    if (root == NULL) return -1;
    if (!root->left && !root->right) {return -1;}
    smin = findSecondMinimumValue(root);
    if (smin == root->val) return -1;
    
    return smin;
}

int findSecondMinimumValue(struct TreeNode* root) {
    int smin;
    
    if (root == NULL || !root->left) {return -1;}
    if (root->val == root->left->val) {
        // root is from left subtree
        smin =  findSecondMinimumValue(root->left);
        if (smin == -1 || smin > root->right->val) {return root->right->val;}
        else {return smin;}
    } else {
        smin =  findSecondMinimumValue(root->right);
        if (smin == -1 || smin > root->left->val) {return root->left->val;}
        else {return smin;}
    }
}

File: Tree/Java/SecondMin.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/**
@Leetcode - https://leetcode.com/problems/second-minimum-node-in-a-binary-tree/
@Youtube - https://www.youtube.com/watch?v=8m_VaMHSCko
@Note - we had to use a boolean flag 'found' to handle a special case of node value = Integer.MAX_VALUE
 */
class Solution {
    int minV;
    int result;
    boolean found;
    
    public int findSecondMinimumValue(TreeNode root) {
        if (root == null) {
            return -1;
        }
        minV = root.val;
        result = Integer.MAX_VALUE;
        found = false;
        
        helper(root);
        return found ? result : -1;
    }
    
    private void helper(TreeNode root) {
        if (root == null) {
            return;
        }
        if (root.val == minV) {
            helper(root.left);
            helper(root.right);
        } else if (root.val > minV && root.val <= result) {
            found = true;
            result = root.val;
        }
    }
}
